syne_tune/optimizer/schedulers/transfer_learning/__init__.py

```diff
@@ -0,0 +1,67 @@
+import pandas as pd
+import numpy as np
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+
+from syne_tune.optimizer.scheduler import TrialScheduler
+
+
+@dataclass
+class TransferLearningTaskEvaluations:
+    """Class that contains offline evaluations for a task that can be used for transfer learning.
+    Args:
+        configuration_space: Dict the configuration space that was used when sampling evaluations.
+        hyperparameters: pd.DataFrame the hyperparameters values that were acquired, all keys of configuration-space
+         should appear as columns.
+        objectives_names: Optional[List[str]] the name of the objectives that were acquired
+        objectives_evaluations: np.array values of recorded objectives, must have shape
+            (num_evals, num_seeds, num_fidelities, num_objectives)
+    """
+    configuration_space: Dict
+    hyperparameters: pd.DataFrame
+    objectives_names: Optional[List[str]]
+    objectives_evaluations: np.array
+
+    def __post_init__(self):
+        assert len(self.objectives_names) == self.objectives_evaluations.shape[-1]
+        assert len(self.hyperparameters) == self.objectives_evaluations.shape[0]
+        for col in self.hyperparameters.keys():
+            assert col in self.configuration_space
+
+    def objective_values(self, objective_name: str) -> np.array:
+        return self.objectives_evaluations[..., self.objective_index(objective_name=objective_name)]
+
+    def objective_index(self, objective_name: str) -> int:
+        matches = [i for i, name in enumerate(self.objectives_names) if name == objective_name]
+        assert len(matches) >= 1, f"could not find objective {objective_name} in recorded objectives " \
+                                  f"{self.objectives_names}"
+        return matches[0]
+
+
+class TransferLearningScheduler(TrialScheduler):
```

Comment:
One idea could be that a "transfer learning scheduler" is always more a factory which takes transfer learning data and some info and then outputs a scheduler (which could still contain the data, but does not have to).

This would allow for many simple things to be done, such as David's bounding box example, where it is about changing the config_space, or also some clever way to construct first configs to try, which would be all about setting the points_to_evaluate argument.

The advantage would be that we can endow existing schedulers with TL capabilities. On the other hand, there will be schedulers that only make sense for TL, but they could also be created by such factories.

Reply:
TransferLearningScheduler provides the following functionalities: 1) it unifies how you would construct a scheduler with access to transfer learning data 2) it performs basic consistency checks. 

Regarding the use of a mixin, we are currently not using them and I am mixed about introducing them. They have pros and cons and if we introduce them here, there are also a couple of other places where we may want to add them (we remove mixin recently due to some issues).

I will reach out to discuss this point directly.

Reply:
@mseeger : just seeing your message now for some reason. As stated above, factories have different trade-offs and having a class does not prevent us from also having a factory. The functionality that you are referring to - being able to wrap any scheduler - is currently supported (I ran experience combining RS, BO and HB with this).